form-test-tom/src/App.jsx

```diff
@@ -0,0 +1,105 @@
+import {useForm} from './useForm'
+import React from 'react'
+import params from './formParams.js'
+
+const animaux = ['bear','tiger','snake','donkey']
+
+const TorchboxForm = () => {
+  const { handleSubmit, handleChange, data, errors } = useForm(params)
+ 
+  return (
+      <form method="post" action="" onSubmit={handleSubmit}>
+        <h1>Fill out this form</h1>
+        <fieldset>
+          <h2>Your details</h2>
+
+          {errors.email && <p className='error'>{errors.email}</p>}
+          <p className={errors.email ? 'error' : ''}>
+            <label className="label" htmlFor="email">Email</label>
+            <input 
+              autoFocus 
+              autoComplete="off"
+              placeholder='you@email.com'
+              type="text" 
+              id="email" 
+              name="email"
+              onChange={handleChange('email')}
+            />
+          </p>
+          
+          {errors.password && <p className='error'>{errors.password}</p>}
+          <p className={errors.password ? 'error' : ''}>
+            <label className="label" htmlFor="password">Password</label>
+            <input
+              type="password"
+              id="password"
+              autoComplete="off"
+              name="username"
+              placeholder="minimum of 8 characters"
+              onChange={handleChange('password')}
+            />
+          </p>
+        </fieldset>
+
+        <fieldset>
+          <h2>Your animals</h2>
+          
+          {errors.colour && <p className='error'>{errors.colour}</p>}
+          <p className={errors.colour ? 'error' : ''}>
+            <label className="label" htmlFor="colour">Colour</label>
+            <select name="colour" id="colour"  onChange={handleChange('colour')}>
+              <option value="">Choose colour</option>
+              <option value="blue">Blue</option>
+              <option value="green">Green</option>
+              <option value="red">Red</option>
+              <option value="black">Black</option>
+              <option value="brown">Brown</option>
+            </select>
+          </p>
+
+          {errors.animals && <p className='error'>{errors.animals}</p>}
+          <p className={errors.animals ? 'error' : ''}>
+            <label className="label">Species</label>
+
+            {animaux.map(animal=><React.Fragment key={animal}>
+              <input 
+                type="checkbox" 
+                name="animal" 
+                value={animal} 
+                id={animal}
+                onChange={handleChange('animals')}
+              />
+              <label htmlFor={animal}>{animal.toUpperCase()}</label>
+            </React.Fragment>)}
+          </p>
+          
+          {errors.tigerType && <p className='error'>{errors.tigerType}</p>}
```

Comment:
by this point, the conditionally-rendering `<p>` could use some DRY love. 
Perhaps any abstracted components could be decorated with a `@withValidationErrorStyling` or similar.